# offline analysis 

by amenallah salem¹ & mohamed amine gaidi²

M2IASD paris-dauphine university

¹ amenallah.salem@dauphine.tn

² Mohamedamine.gaidi@dauphine.tn

In [1]:
#if you are runing this notebook on google colab, just run this cell 
#and upload the two datasets on google drive 
#then mount the drive to the data path 
####################################################################################
#from google.colab import drive
#drive.mount('/content/drive')
####################################################################################
#if not (runing it on your jupyter notebook on pc) 
#just set the wright drictory to the data set path 
#

In [2]:
#importing packages 
import csv
import pandas as pd
from pandas import DataFrame
import matplotlib.pyplot as plt 
%matplotlib inline 
import seaborn as sns

In [3]:
data_path='/home/amenallah/Desktop/flux de donnee/projet/offline_analysis/'
clickStream_df = pd.read_json(data_path+'clickStream.json', lines=True)
clickStream_df.name="clickStream data"
displayStream_df = pd.read_json(data_path+'displayStream.json', lines=True)
displayStream_df.name='displayStream data'

In [4]:
clickStream_df.head()

eventType                                     uid           timestamp  \
0     click    d1dcffd7-6766-4416-8d14-c754f77b15ec 2020-05-11 22:40:20   
1     click    aa5eeb1b-89a5-473f-91c5-e089c51adfee 2020-05-11 22:40:20   
2     click    3acda4d7-8a1e-4741-8372-605457661ffc 2020-05-11 22:40:20   
3     click  279ddb6c-03f1-4474-9b08-66dfb5881a4e10 2020-05-11 22:40:21   
4     click  9864d3e5-496f-4015-be9c-430b8a4a4ad511 2020-05-11 22:40:21   

                ip                          impressionId  
0      4.61.175.17  b327f163-f47f-4693-88a6-817e33977c78  
1     188.36.54.15  cf821c1c-5735-4c0a-914f-766f8372f11c  
2  157.140.250.176  6167acae-c894-4918-b013-031b6589d8d7  
3    238.186.83.58  fe7c3093-640c-4a8b-b815-0f5a53ff5a36  
4    238.186.83.58  2f9547ca-7059-4a65-80de-6631b2252805

In [5]:
print("##############################################################")
print("Some information about clickStream_df")
print("##############################################################")
print("The shape of the clickStream_df data is ")
print(clickStream_df.shape)
print("##############################################################")
print("other useful information")
print(clickStream_df.info())
print("##############################################################")
print("checking of missing values")
print(clickStream_df.isnull().sum())

##############################################################
Some information about clickStream_df
##############################################################
The shape of the clickStream_df data is 
(3901, 5)
##############################################################
other useful information
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3901 entries, 0 to 3900
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   eventType     3901 non-null   object        
 1   uid           3901 non-null   object        
 2   timestamp     3901 non-null   datetime64[ns]
 3   ip            3901 non-null   object        
 4   impressionId  3901 non-null   object        
dtypes: datetime64[ns](1), object(4)
memory usage: 152.5+ KB
None
##############################################################
checking of missing values
eventType       0
uid             0
timestamp       0
ip              0
impressionId 

In [6]:
displayStream_df

eventType                                   uid           timestamp  \
0       display  0d1c091f-e704-4e6f-bb98-7ee5ff6a70f6 2020-05-11 22:56:21   
1       display  79a04aa3-3245-4142-b4a7-ccfca38ab9be 2020-05-11 22:56:21   
2       display  6b49cf0d-f161-43f4-9b91-068ab05c94d1 2020-05-11 22:56:21   
3       display  af6db700-bd33-477e-830f-031a0cccfca1 2020-05-11 22:56:21   
4       display  070e1f39-f091-468f-889d-40af9fa15225 2020-05-11 22:56:21   
...         ...                                   ...                 ...   
28179   display  8365f279-c7bf-48ff-beb5-d9d35dbbb8aa 2020-05-11 23:29:38   
28180   display  c6cb5f2d-a10f-4a76-913a-0bd1b94125de 2020-05-11 23:29:38   
28181   display  3452541c-64c6-40a1-af6d-099dd2919b8b 2020-05-11 23:29:38   
28182   display  94292770-6164-43b4-a695-9be9d6854c05 2020-05-11 23:29:38   
28183   display  34fc0552-faf0-47bd-9d71-9b100803535c 2020-05-11 23:29:38   

                    ip                          impressionId  
0      163.241.120.182  426cd1c9-e299-4b91-8ead-769b58d6b300  
1        106.252.112.1  0e124861-ecdf-494d-bbcc-069a13ebe2a6  
2       31.108.101.220  aca5e651-2a63-4fc9-9322-5aa2c5f17b86  
3        22.105.66.149  d8e191b9-88d2-4e03-8f89-61dc9a382ab0  
4         214.47.42.50  96b812dc-00ee-4af3-a851-524eb995f4a5  
...                ...                                   ...  
28179     167.105.1.61  269afa6f-7465-483a-a92e-87a4cff6f35a  
28180   12.157.138.101  28a39bbe-2b65-468e-bff1-89b8fec7b006  
28181  153.139.252.121  8a9fe72a-064f-455f-bac3-94df586a962c  
28182     199.42.34.79  677db765-a1fe-4858-9546-fd59ca8df2bc  
28183      205.6.84.92  f466448d-3556-4a98-9b7a-b5ff7d4e90b5  

[28184 rows x 5 columns]

In [7]:
print("##############################################################")
print("Some information about displayStream_df")
print("##############################################################")
print("The shape of the displayStream_df data is ")
print(displayStream_df.shape)
print("##############################################################")
print("other useful information")
print(displayStream_df.info())
print("##############################################################")
print("checking of missing values")
print(displayStream_df.isnull().sum())

##############################################################
Some information about displayStream_df
##############################################################
The shape of the displayStream_df data is 
(28184, 5)
##############################################################
other useful information
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28184 entries, 0 to 28183
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   eventType     28184 non-null  object        
 1   uid           28184 non-null  object        
 2   timestamp     28184 non-null  datetime64[ns]
 3   ip            28184 non-null  object        
 4   impressionId  28184 non-null  object        
dtypes: datetime64[ns](1), object(4)
memory usage: 1.1+ MB
None
##############################################################
checking of missing values
eventType       0
uid             0
timestamp       0
ip              0
impressi

# Let's detect our patern
## working with clickStream_df



First method direct using the fact thet the  uid must be ==ip adress
    
**clickStream_df preprocessing**

In [8]:
def clickStream_df_preprocessing(df_):
    #we first drop the duplicated rows based on their uid
    df_=df_.drop_duplicates(subset = ["uid"], keep ="first")
    #put all unique ip adresses in a np array (using the unique method)!
    uniqueIp = df_['ip'].unique()
    #display uniques ip adresses 
    print('the unisue IP adresses in our dataset are ', uniqueIp)
    print('********************************************************')
    print('the data set cleaned')
    return df_


In [12]:
clickStream_df_preprocessing(clickStream_df)

the unisue IP adresses in our dataset are  ['4.61.175.17' '188.36.54.15' '157.140.250.176' '238.186.83.58'
 '29.15.119.214' '204.179.119.30' '247.215.74.212' '146.197.162.167'
 '202.40.143.123' '132.212.131.110' '59.44.123.242' '195.48.17.43'
 '143.43.209.232' '89.64.214.112' '174.172.15.221' '44.198.43.221'
 '10.140.9.233' '94.91.252.57' '189.167.131.96' '110.85.188.110'
 '63.191.90.199' '79.245.208.0' '140.189.218.212' '213.220.103.78'
 '236.58.238.21' '65.93.153.29' '238.28.107.158' '24.56.161.195'
 '10.186.221.249' '215.238.73.150' '58.70.111.12' '103.243.42.143'
 '234.184.93.117' '135.48.34.187' '238.90.241.253' '169.87.81.191'
 '246.60.209.233' '20.138.110.246' '170.92.175.134' '14.47.58.4'
 '254.255.232.95' '250.159.156.178' '82.25.225.14' '35.157.205.199'
 '133.15.41.208' '235.172.130.36' '198.253.250.106' '133.47.58.89'
 '155.172.186.223' '201.135.21.50' '217.179.224.211' '46.243.159.133'
 '86.95.67.66' '229.177.27.106' '89.21.69.145' '134.142.59.247'
 '238.150.202.11' '235.55

eventType                                     uid           timestamp  \
0        click    d1dcffd7-6766-4416-8d14-c754f77b15ec 2020-05-11 22:40:20   
1        click    aa5eeb1b-89a5-473f-91c5-e089c51adfee 2020-05-11 22:40:20   
2        click    3acda4d7-8a1e-4741-8372-605457661ffc 2020-05-11 22:40:20   
3        click  279ddb6c-03f1-4474-9b08-66dfb5881a4e10 2020-05-11 22:40:21   
4        click  9864d3e5-496f-4015-be9c-430b8a4a4ad511 2020-05-11 22:40:21   
...        ...                                     ...                 ...   
3896     click  d44e93dc-3ade-4d79-8616-5b8f5991006f15 2020-05-11 22:53:51   
3897     click  5d6df97e-7c38-43d6-b116-a6202192b2e616 2020-05-11 22:53:51   
3898     click  15a7f68d-3460-454e-af8e-05b0bd8ca42217 2020-05-11 22:53:51   
3899     click  2104dc08-7ec9-42c2-8d06-c85beb15a7ed18 2020-05-11 22:53:51   
3900     click  9704d9fa-1f0d-4e6c-ac83-3d70a700e30e19 2020-05-11 22:53:51   

                   ip                          impressionId  
0         4.61.175.17  b327f163-f47f-4693-88a6-817e33977c78  
1        188.36.54.15  cf821c1c-5735-4c0a-914f-766f8372f11c  
2     157.140.250.176  6167acae-c894-4918-b013-031b6589d8d7  
3       238.186.83.58  fe7c3093-640c-4a8b-b815-0f5a53ff5a36  
4       238.186.83.58  2f9547ca-7059-4a65-80de-6631b2252805  
...               ...                                   ...  
3896    238.186.83.58  328b7acb-a898-48e3-8075-ecf263f66563  
3897    238.186.83.58  794d8d2c-25dc-4079-9608-51a77732d09a  
3898    238.186.83.58  e0648864-485d-459d-a43c-c4071bdd15b6  
3899    238.186.83.58  85778c4f-0c80-4b68-ade4-3f51959457a7  
3900    238.186.83.58  2aae3c31-d586-4540-b040-e0ed8445b4e9  

[1577 rows x 5 columns]

In [13]:
def outlier_detection(df_):
    if len(df_)==len(df_['ip'].unique()):
        print("the unique Ip adresses is egual to unique uid's")
    else: 
        print("there are {} unique Ip within {} uid's rows in the {}, some IP's are repeated.".format(len(df_['ip'].unique()),len(df_), df_.name ) )



outlier_detection(clickStream_df)

there are 2628 unique Ip within 3901 uid's rows in the clickStream data, some IP's are repeated.


# working with displayStream_df

The same thing for the displayStream_df

In [14]:
def displayStream_df_preprocessing(df_):
    #we first drop the duplicated rows based on their uid
    df_=df_.drop_duplicates(subset = ["uid"], keep ="first")
    #put all unique ip adresses in a np array (using the unique method)!
    uniqueDispIp = df_['ip'].unique()
    #display uniques ip adresses 
    print('the unisue IP adresses in our dataset are ', uniqueDispIp)
    print('********************************************************')
    print('the data set cleaned')
    return df_

In [15]:
displayStream_df_preprocessing(displayStream_df)

the unisue IP adresses in our dataset are  ['163.241.120.182' '106.252.112.1' '31.108.101.220' ... '205.239.140.120'
 '93.227.188.204' '199.42.34.79']
********************************************************
the data set cleaned


eventType                                   uid           timestamp  \
0       display  0d1c091f-e704-4e6f-bb98-7ee5ff6a70f6 2020-05-11 22:56:21   
1       display  79a04aa3-3245-4142-b4a7-ccfca38ab9be 2020-05-11 22:56:21   
2       display  6b49cf0d-f161-43f4-9b91-068ab05c94d1 2020-05-11 22:56:21   
3       display  af6db700-bd33-477e-830f-031a0cccfca1 2020-05-11 22:56:21   
4       display  070e1f39-f091-468f-889d-40af9fa15225 2020-05-11 22:56:21   
...         ...                                   ...                 ...   
28174   display  f4521a51-e564-48f0-af93-250fa90c612d 2020-05-11 23:29:38   
28176   display  72a7cd4e-514b-4990-8c93-6c224fab10a7 2020-05-11 23:29:38   
28178   display  daaeeeaa-685b-481b-aa78-b379a1518ad1 2020-05-11 23:29:38   
28180   display  c6cb5f2d-a10f-4a76-913a-0bd1b94125de 2020-05-11 23:29:38   
28182   display  94292770-6164-43b4-a695-9be9d6854c05 2020-05-11 23:29:38   

                    ip                          impressionId  
0      163.241.120.182  426cd1c9-e299-4b91-8ead-769b58d6b300  
1        106.252.112.1  0e124861-ecdf-494d-bbcc-069a13ebe2a6  
2       31.108.101.220  aca5e651-2a63-4fc9-9322-5aa2c5f17b86  
3        22.105.66.149  d8e191b9-88d2-4e03-8f89-61dc9a382ab0  
4         214.47.42.50  96b812dc-00ee-4af3-a851-524eb995f4a5  
...                ...                                   ...  
28174     37.93.74.174  801b813d-26f9-40e0-bfed-361241d75623  
28176    78.51.100.171  5095d095-2cbe-4ac5-981c-e81b2ac635d6  
28178    11.249.97.157  7ee41ee4-f155-4d5f-bb1b-ff6689dcc256  
28180   12.157.138.101  28a39bbe-2b65-468e-bff1-89b8fec7b006  
28182     199.42.34.79  677db765-a1fe-4858-9546-fd59ca8df2bc  

[14973 rows x 5 columns]

In [16]:
outlier_detection(displayStream_df)

there are 9269 unique Ip within 28184 uid's rows in the displayStream data, some IP's are repeated.


# conclusion 

Likewise, as there are 7258 unique IP within 14973 uid's rows in the displayStream, some IP's are repeated. \\
**THIS IS LIKELY TO BE A FRAUD!!**